compiler/cli/src/org/jetbrains/kotlin/cli/jklib/K2JKlibCompiler.kt

```diff
@@ -0,0 +1,948 @@
+
+        private fun DeclarationDescriptor.isJavaDescriptor(): Boolean {
+            if (this is PackageFragmentDescriptor) {
+                return this is LazyJavaPackageFragment || fqName.startsWith(javaName)
+            }
+
+            return this is JavaClassDescriptor || this is JavaCallableMemberDescriptor || (containingDeclaration?.isJavaDescriptor() == true)
+        }
+
+        override fun platformSpecificSymbol(symbol: IrSymbol): Boolean {
+            return symbol.descriptor.isJavaDescriptor()
+        }
+
+        private fun DeclarationDescriptor.isCleanDescriptor(): Boolean {
+            if (this is PropertyAccessorDescriptor) return correspondingProperty.isCleanDescriptor()
+            return this is DeserializedDescriptor
+        }
+
+        private fun declareJavaFieldStub(symbol: IrFieldSymbol): IrField {
+            return with(stubGenerator) {
+                val old = stubGenerator.unboundSymbolGeneration
+                try {
+                    stubGenerator.unboundSymbolGeneration = true
+                    generateFieldStub(symbol.descriptor)
+                } finally {
+                    stubGenerator.unboundSymbolGeneration = old
+                }
+            }
+        }
+
+        override val fakeOverrideBuilder = IrLinkerFakeOverrideProvider(
+            linker = this,
+            symbolTable = symbolTable,
+            mangler = JKlibIrMangler,
+            typeSystem = IrTypeSystemContextImpl(builtIns),
+            friendModules = emptyMap(),
+            partialLinkageSupport = partialLinkageSupport,
+        )
+
+        override fun isBuiltInModule(moduleDescriptor: ModuleDescriptor): Boolean =
+            moduleDescriptor === moduleDescriptor.builtIns.builtInsModule
+
+        private val IrLibrary.libContainsErrorCode: Boolean
+            get() = this is KotlinLibrary && this.containsErrorCode
+
+        override fun createModuleDeserializer(
+            moduleDescriptor: ModuleDescriptor,
+            klib: KotlinLibrary?,
+            strategyResolver: (String) -> DeserializationStrategy,
+        ): IrModuleDeserializer {
+            if (klib == null) {
+                return MetadataJVMModuleDeserializer(moduleDescriptor, emptyList())
+            }
+
+            val libraryAbiVersion = klib.versions.abiVersion ?: KotlinAbiVersion.CURRENT
+            return JKlibModuleDeserializer(
+                moduleDescriptor,
+                klib,
+                strategyResolver,
+                libraryAbiVersion,
+                klib.libContainsErrorCode,
+            )
+        }
+
+        private val mappedClassSymbols = mutableMapOf<FqName, IrClassSymbol>()
+
+        private fun FqName.isMappedType(): Boolean {
+            return (JavaToKotlinClassMap.mapJavaToKotlin(this) ?: JavaToKotlinClassMap.mapKotlinToJava(toUnsafe())) != null
+        }
+
+        private fun withKotlinBuiltinsHack(idSig: IdSignature, f: () -> IrSymbol?): IrSymbol? {
```

Comment:
This is the workaround for mapped types that I mentioned in the description of this PR